### PR TITLE
Add a magician command for reassigning reviewers

### PR DESCRIPTION
**Pull request by trodge**
This command will allow reassigning a reviewer either to a random core contributor or to a specific one.

Currently can be invoked with
```
cd .ci/magician
go build .
./magician reassign-reviewer PR_NUMBER [REVIEWER]
```
(REVIEWER is optional, omitting will assign to a random core reviewer)

helps with https://github.com/hashicorp/terraform-provider-google/issues/19691

Remaining TODOs:
* Add a workflow that runs `reassign-reviewer` on `@modular-magician reassign-reviewer [reviewer]`
* Make the re-request reviewer workflow check the current reviewer using `FindReviewerComment`

**Release Note Template for Downstream PRs (will be copied)**

```release-note:none

```

### CODE REVIEW

File: .ci/magician/github/reviewer_assignment.go
Comment (c2thorn) [inline .ci/magician/github/reviewer_assignment.go:77]: what order are the comments returned in? first -> last?

If so, should it be reversed so that the latest reviewer comment is used, and not the first

**Review by c2thorn — APPROVED** [review]
it'd be great to get this into a PR-command like gcbrun

LGTM, left minor comment about reviewer comment order

File: .ci/magician/github/reviewer_assignment.go
Comment (trodge) [reply .ci/magician/github/reviewer_assignment.go:77]: Not sure if it's returned in any particular order. Probably the best thing would be to compare the creation times.

Comment (c2thorn) [reply .ci/magician/github/reviewer_assignment.go:77]: Just minor detail, small chance the user may have to run the command twice if there's been multiple reviewers

**Comment by trodge** [comment]
> it'd be great to get this into a PR-command like gcbrun
> 
> LGTM, left minor comment about reviewer comment order

The plan is to use a workflow that triggers on PR comments to invoke this command.

File: .ci/magician/cmd/reassign_reviewer.go
Comment (melinath) [inline .ci/magician/cmd/reassign_reviewer.go:46]: This shouldn't need the fallback

File: .ci/magician/cmd/reassign_reviewer.go
Comment (melinath) [inline .ci/magician/cmd/reassign_reviewer.go:60]:
```suggestion
	comments, err := gh.GetPullRequestCommentsByUser(prNumber, "github-actions")
```

File: .ci/magician/github/get.go
Comment (melinath) [inline .ci/magician/github/get.go:118]: Are we expecting to reuse this? What would you think about just keeping it to GetPullRequestComments (which will be easier to port over to the standard GH client once we get around to that conversion) and handling the rest as part of FindReviewerComment?

File: .ci/magician/github/reviewer_assignment.go
Comment (melinath) [inline .ci/magician/github/reviewer_assignment.go:71]: We should definitely have a test that verifies that this matches the output of the function that formats this comment.

File: .ci/magician/github/reviewer_assignment.go
Comment (melinath) [reply .ci/magician/github/reviewer_assignment.go:77]: @c2thorn there should only ever be one comment that contains the reviewer - the proposed implementation is to update that comment when the reviewer is reassigned (so that we're not creating more noise in the PR). I don't _think_ there's a way for us to end up with more than one, but FWIW if it ends up being a common issue I'd prefer to delete or hide the duplicates.

**Review by melinath — CHANGES_REQUESTED** [review]

File: .ci/magician/cmd/reassign_reviewer.go
Comment (trodge) [reply .ci/magician/cmd/reassign_reviewer.go]: Going to remove the filter altogether because the comment could come from modular magician if we have to make a new comment because none was found (as is the case for PRs from core contributors).

File: .ci/magician/github/get.go
Comment (trodge) [reply .ci/magician/github/get.go]: Removing this altogether since reviewer comment doesn't necessarily come from a specific user.

File: .ci/magician/github/reviewer_assignment.go
Comment (trodge) [reply .ci/magician/github/reviewer_assignment.go:71]: Done

File: .ci/magician/github/reviewer_assignment.go
Comment (trodge) [reply .ci/magician/github/reviewer_assignment.go:79]: Just in case, I've made it return the most recent reviewer comment.

File: .ci/magician/cmd/reassign_reviewer.go
Comment (melinath) [reply .ci/magician/cmd/reassign_reviewer.go]: FWIW I believe that (since this is going to run in a GHA) the comment will always come from `github-actions`. But skipping the check is also probably fine.

File: .ci/magician/cmd/reassign_reviewer.go
Comment (melinath) [inline .ci/magician/cmd/reassign_reviewer.go:47]: I think this should just use GITHUB_TOKEN since it'll run in the context of a GHA.

```suggestion
		githubToken, ok := os.LookupEnv("GITHUB_TOKEN")
```

File: .ci/magician/cmd/reassign_reviewer.go
Comment (melinath) [inline .ci/magician/cmd/reassign_reviewer.go:91]: nit: this kind of logging should happen in the caller, since that's where the decision about why this was called was made.

File: .ci/magician/github/reviewer_assignment.go
Comment (melinath) [reply .ci/magician/github/reviewer_assignment.go:79]: IMO It should take the oldest reviewer comment rather than the most recent. But since there should only be one it's not a blocker for me at this time.

File: .ci/magician/github/reviewer_assignment.go
Comment (melinath) [inline .ci/magician/github/reviewer_assignment.go:96]: I think this could be simplified a bit.

```suggestion
		matches := reviewerCommentRegex.FindStringSubmatch(comment.Body)
		fmt.Println(comment, matches)
		if matches == nil {
			// Skip comments that don't match regex.
			continue
		}
		newestComment = comment
		currentReviewer = matches[1]
```

**Review by melinath — APPROVED** [review]

File: .ci/magician/cmd/reassign_reviewer.go
Comment (trodge) [reply .ci/magician/cmd/reassign_reviewer.go:47]: Done

File: .ci/magician/github/reviewer_assignment.go
Comment (melinath) [reply .ci/magician/github/reviewer_assignment.go]: I hear you, and I wish golang provided an easier way to access the matches by group name, but this implementation is currently a bit convoluted considering there's only one subgroup in the regex. It would be a bit more clear if you did something like [make a map of all the named groups and then extracted the value from that](https://stackoverflow.com/questions/20750843/using-named-matches-from-go-regex) but that also feels pretty heavy for what's happening here.

Comment (melinath) [reply .ci/magician/github/reviewer_assignment.go]: We could factor the map logic out into a separate util library for handling regex but this particular case doesn't provide a strong justification for it.

Comment (melinath) [reply .ci/magician/github/reviewer_assignment.go]: Just to close this discussion out - I'm not blocking the merge based on this.

**Pull request merged**
